File: internal/tsworker/framing.go
// Package tsworker implements the Go side of the Luque TypeScript worker
// protocol described in docs/protocol/ts-worker-v1.md.
package tsworker

import (
	"context"
	"encoding/binary"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"sync"
	"time"
)

const (
	// ProtocolVersion is the only protocol version this package speaks.
	ProtocolVersion = 1
	// MaxFrameBytes is the largest body accepted or produced, per the protocol.
	MaxFrameBytes = 16 << 20
	// frameHeaderBytes is the size of the big-endian length prefix.
	frameHeaderBytes = 4
)

// FramingErrorKind classifies transport failures. Fatal kinds end the session
// because the byte stream can no longer be trusted; the protocol forbids
// resynchronising a corrupted stream.
type FramingErrorKind string

const (
	// FrameTooLarge reports a length prefix above MaxFrameBytes. Fatal.
	FrameTooLarge FramingErrorKind = "FRAME_TOO_LARGE"
	// FrameEmpty reports a zero-length body, which the protocol forbids. Fatal.
	FrameEmpty FramingErrorKind = "FRAME_EMPTY"
	// FrameTruncated reports end of input in the middle of a frame. Fatal.
	FrameTruncated FramingErrorKind = "FRAME_TRUNCATED"
	// InvalidPayload reports a well-read frame whose body is not a valid
	// envelope. Recoverable: the stream stays aligned.
	InvalidPayload FramingErrorKind = "INVALID_PAYLOAD"
	// VersionMismatch reports an envelope declaring another protocol version. Fatal.
	VersionMismatch FramingErrorKind = "VERSION_MISMATCH"
	// Timeout reports a read or write that exceeded its deadline. Fatal.
	Timeout FramingErrorKind = "TIMEOUT"
	// Canceled reports a read or write stopped by context cancellation. Fatal.
	Canceled FramingErrorKind = "CANCELED"
	// IOFailure reports an underlying transport failure. Fatal.
	IOFailure FramingErrorKind = "IO_FAILURE"
)

// ErrSessionClosed reports use of a reader or writer after Close.
var ErrSessionClosed = errors.New("tsworker session is closed")

// FramingError is a classified transport failure.
type FramingError struct {
	Kind FramingErrorKind
	Op   string
	Err  error
}

func (err *FramingError) Error() string {
	if err == nil {
		return "<nil>"
	}
	if err.Err == nil {
		return fmt.Sprintf("tsworker %s: %s", err.Op, err.Kind)
	}
	return fmt.Sprintf("tsworker %s: %s: %v", err.Op, err.Kind, err.Err)
}

func (err *FramingError) Unwrap() error {
	if err == nil {
		return nil
	}
	return err.Err
}

// Fatal reports whether the error invalidates the session. Only InvalidPayload
// leaves the stream usable, because the frame boundary was still honoured.
func (err *FramingError) Fatal() bool {
	return err != nil && err.Kind != InvalidPayload
}

func newFramingError(kind FramingErrorKind, op string, cause error) error {
	return &FramingError{Kind: kind, Op: op, Err: cause}
}

// Envelope is the outer object carried by every frame.
type Envelope struct {
	V       int             `json:"v"`
	ID      uint64          `json:"id"`
	Type    string          `json:"type"`
	Payload json.RawMessage `json:"payload"`
}

// Validate checks the invariants the protocol requires of every envelope.
func (envelope Envelope) Validate() error {
	if envelope.V != ProtocolVersion {
		return fmt.Errorf("protocol version %d is not supported", envelope.V)
	}
	if envelope.Type == "" {
		return errors.New("message type must not be empty")
	}
	if len(envelope.Payload) == 0 {
		return errors.New("payload must be present")
	}
	return nil
}

// deadlineSetter is implemented by *os.File pipes and net.Conn, which is how
// a blocking read is actually interrupted. A transport without it can only be
// cancelled between frames.
type deadlineSetter interface {
	SetReadDeadline(time.Time) error
}

// Reader decodes frames from a transport. It is not safe for concurrent use.
type Reader struct {
	source   io.Reader
	deadline deadlineSetter
	header   [frameHeaderBytes]byte
	body     []byte
	closed   bool
}

// NewReader builds a frame reader. When source supports read deadlines, an
// in-flight read is interrupted on context cancellation; otherwise the context
// is only observed between frames.
func NewReader(source io.Reader) *Reader {
	reader := &Reader{source: source}
	if candidate, ok := source.(deadlineSetter); ok {
		reader.deadline = candidate
	}
	return reader
}

// SupportsInterruption reports whether a blocked read can be cancelled.
func (reader *Reader) SupportsInterruption() bool {
	return reader != nil && reader.deadline != nil
}

// Close releases the reader. Further reads fail with ErrSessionClosed.
func (reader *Reader) Close() {
	if reader != nil {
		reader.closed = true
		reader.body = nil
	}
}

// ReadFrame reads exactly one frame and decodes its envelope. The returned
// payload slice is owned by the reader and is reused by the next call.
func (reader *Reader) ReadFrame(ctx context.Context) (Envelope, error) {
	if reader == nil || reader.closed {
		return Envelope{}, newFramingError(IOFailure, "read", ErrSessionClosed)
	}
	if ctx == nil {
		ctx = context.Background()
	}
	if err := ctx.Err(); err != nil {
		return Envelope{}, newFramingError(Canceled, "read", err)
	}

	stop := reader.armDeadline(ctx)
	defer stop()

	if _, err := io.ReadFull(reader.source, reader.header[:]); err != nil {
		return Envelope{}, reader.classify(ctx, "read header", err, true)
	}
	length := binary.BigEndian.Uint32(reader.header[:])
	if length == 0 {
		return Envelope{}, newFramingError(FrameEmpty, "read header", nil)
	}
	// The length is validated before any allocation, so a hostile prefix cannot
	// make the process reserve 4 GiB.
	if length > MaxFrameBytes {
		return Envelope{}, newFramingError(FrameTooLarge, "read header", fmt.Errorf("length %d exceeds %d", length, MaxFrameBytes))
	}

	if uint32(cap(reader.body)) < length {
		reader.body = make([]byte, length)
	}
	body := reader.body[:length]
	if _, err := io.ReadFull(reader.source, body); err != nil {
		return Envelope{}, reader.classify(ctx, "read body", err, false)
	}

	var envelope Envelope
	if err := json.Unmarshal(body, &envelope); err != nil {
		return Envelope{}, newFramingError(InvalidPayload, "decode", err)
	}
	if envelope.V != ProtocolVersion {
		return Envelope{}, newFramingError(VersionMismatch, "decode", fmt.Errorf("protocol version %d is not supported", envelope.V))
	}
	if err := envelope.Validate(); err != nil {
		return Envelope{}, newFramingError(InvalidPayload, "decode", err)
	}
	return envelope, nil
}

// armDeadline interrupts a blocked read when the context ends, and restores the
// transport afterwards.
func (reader *Reader) armDeadline(ctx context.Context) func() {
	if reader.deadline == nil || ctx.Done() == nil {
		return func() {}
	}
	if deadline, ok := ctx.Deadline(); ok {
		_ = reader.deadline.SetReadDeadline(deadline)
	}
	finished := make(chan struct{})
	watcherDone := make(chan struct{})
	go func() {
		defer close(watcherDone)
		select {
		case <-ctx.Done():
			_ = reader.deadline.SetReadDeadline(time.Unix(1, 0))
		case <-finished:
		}
	}()
	return func() {
		// The watcher must be observed as finished before the deadline is
		// cleared; otherwise a late interruption would leak an expired
		// deadline into the next read.
		close(finished)
		<-watcherDone
		_ = reader.deadline.SetReadDeadline(time.Time{})
	}
}

// classify maps a transport failure onto a protocol error. A clean EOF between
// frames ends the session; the same EOF inside a frame is truncation.
func (reader *Reader) classify(ctx context.Context, op string, cause error, atBoundary bool) error {
	if ctxErr := ctx.Err(); ctxErr != nil {
		if errors.Is(ctxErr, context.DeadlineExceeded) {
			return newFramingError(Timeout, op, ctxErr)
		}
		return newFramingError(Canceled, op, ctxErr)
	}
	if errors.Is(cause, io.EOF) {
		if atBoundary {
			return io.EOF
		}
		return newFramingError(FrameTruncated, op, cause)
	}
	if errors.Is(cause, io.ErrUnexpectedEOF) {
		return newFramingError(FrameTruncated, op, cause)
	}
	if isTimeout(cause) {
		return newFramingError(Timeout, op, cause)
	}
	return newFramingError(IOFailure, op, cause)
}

func isTimeout(err error) bool {
	var timeout interface{ Timeout() bool }
	return errors.As(err, &timeout) && timeout.Timeout()
}

// Writer encodes frames onto a transport. It is safe for concurrent use so the
// supervisor can answer while emitting events.
type Writer struct {
	mu     sync.Mutex
	sink   io.Writer
	buffer []byte
	closed bool
}

// NewWriter builds a frame writer.
func NewWriter(sink io.Writer) *Writer {
	return &Writer{sink: sink}
}

// Close releases the writer. Further writes fail with ErrSessionClosed.
func (writer *Writer) Close() {
	if writer == nil {
		return
	}
	writer.mu.Lock()
	writer.closed = true
	writer.buffer = nil
	writer.mu.Unlock()
}

// WriteFrame encodes one envelope as a single frame.
func (writer *Writer) WriteFrame(envelope Envelope) error {
	if writer == nil {
		return newFramingError(IOFailure, "write", ErrSessionClosed)
	}
	if err := envelope.Validate(); err != nil {
		return newFramingError(InvalidPayload, "encode", err)
	}
	body, err := json.Marshal(envelope)
	if err != nil {
		return newFramingError(InvalidPayload, "encode", err)
	}
	if len(body) > MaxFrameBytes {
		return newFramingError(FrameTooLarge, "encode", fmt.Errorf("length %d exceeds %d", len(body), MaxFrameBytes))
	}

	writer.mu.Lock()
	defer writer.mu.Unlock()
	if writer.closed {
		return newFramingError(IOFailure, "write", ErrSessionClosed)
	}
	total := frameHeaderBytes + len(body)
	if cap(writer.buffer) < total {
		writer.buffer = make([]byte, total)
	}
	frame := writer.buffer[:total]
	binary.BigEndian.PutUint32(frame[:frameHeaderBytes], uint32(len(body)))
	copy(frame[frameHeaderBytes:], body)
	if _, err := writer.sink.Write(frame); err != nil {
		if isTimeout(err) {
			return newFramingError(Timeout, "write", err)
		}
		return newFramingError(IOFailure, "write", err)
	}
	return nil
}

// NewEnvelope builds a validated envelope with the current protocol version.
func NewEnvelope(id uint64, messageType string, payload any) (Envelope, error) {
	encoded, err := json.Marshal(payload)
	if err != nil {
		return Envelope{}, newFramingError(InvalidPayload, "encode", err)
	}
	envelope := Envelope{V: ProtocolVersion, ID: id, Type: messageType, Payload: encoded}
	if err := envelope.Validate(); err != nil {
		return Envelope{}, newFramingError(InvalidPayload, "encode", err)
	}
	return envelope, nil
}

File: ts-worker/src/framing.ts
/**
 * Wire codec for the Kivgraph TypeScript worker protocol, version 1.
 *
 * The contract lives in docs/protocol/ts-worker-v1.md and must stay byte for
 * byte compatible with internal/tsworker in Go. The shared fixtures under
 * testdata/protocol/ts-worker-v1 are the executable proof of that compatibility.
 */

/** Only protocol version this codec speaks. */
export const PROTOCOL_VERSION = 1;
/** Largest body accepted or produced, per the protocol. */
export const MAX_FRAME_BYTES = 16 << 20;
/** Size of the big-endian length prefix. */
export const FRAME_HEADER_BYTES = 4;

/**
 * Transport failure classification. Every kind except INVALID_PAYLOAD ends the
 * session, because the byte stream can no longer be trusted; the protocol
 * forbids resynchronising a corrupted stream.
 */
export type FramingErrorKind =
  | "FRAME_TOO_LARGE"
  | "FRAME_EMPTY"
  | "FRAME_TRUNCATED"
  | "INVALID_PAYLOAD"
  | "VERSION_MISMATCH"
  | "IO_FAILURE";

/** Classified transport failure, mirroring the Go FramingError. */
export class FramingError extends Error {
  readonly kind: FramingErrorKind;
  readonly op: string;

  constructor(kind: FramingErrorKind, op: string, detail?: string) {
    super(
      detail ? `tsworker ${op}: ${kind}: ${detail}` : `tsworker ${op}: ${kind}`,
    );
    this.name = "FramingError";
    this.kind = kind;
    this.op = op;
  }

  /** Reports whether the failure invalidates the session. */
  get fatal(): boolean {
    return this.kind !== "INVALID_PAYLOAD";
  }
}

/** Outer object carried by every frame. */
export interface Envelope {
  readonly v: number;
  readonly id: number;
  readonly type: string;
  readonly payload: unknown;
}

function isPlainObject(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

/**
 * Serialises a value the way Go's encoding/json serialises a map: object keys
 * sorted, no insignificant whitespace. Payloads travel as Go maps, so sorting
 * is what reproduces the same bytes on both sides.
 */
export function canonicalJSON(value: unknown): string {
  if (value === null || typeof value !== "object") {
    return JSON.stringify(value) ?? "null";
  }
  if (Array.isArray(value)) {
    return `[${value.map((entry) => canonicalJSON(entry)).join(",")}]`;
  }
  const record = value as Record<string, unknown>;
  const entries = Object.keys(record)
    .sort()
    .filter((key) => record[key] !== undefined)
    .map((key) => `${JSON.stringify(key)}:${canonicalJSON(record[key])}`);
  return `{${entries.join(",")}}`;
}

/** Checks the invariants the protocol requires of every envelope. */
export function validateEnvelope(envelope: Envelope): void {
  if (envelope.v !== PROTOCOL_VERSION) {
    throw new FramingError(
      "INVALID_PAYLOAD",
      "encode",
      `protocol version ${envelope.v} is not supported`,
    );
  }
  if (typeof envelope.type !== "string" || envelope.type === "") {
    throw new FramingError(
      "INVALID_PAYLOAD",
      "encode",
      "message type must not be empty",
    );
  }
  if (!Number.isInteger(envelope.id) || envelope.id < 0) {
    throw new FramingError(
      "INVALID_PAYLOAD",
      "encode",
      "id must be a non-negative integer",
    );
  }
  if (envelope.payload === undefined || envelope.payload === null) {
    throw new FramingError(
      "INVALID_PAYLOAD",
      "encode",
      "payload must be present",
    );
  }
}

/** Builds a validated envelope with the current protocol version. */
export function newEnvelope(
  id: number,
  type: string,
  payload: unknown,
): Envelope {
  const envelope: Envelope = { v: PROTOCOL_VERSION, id, type, payload };
  validateEnvelope(envelope);
  return envelope;
}

/**
 * Encodes one envelope as a complete frame. Envelope keys are emitted in the
 * order the Go struct declares them, and the payload uses canonical ordering.
 */
export function encodeFrame(envelope: Envelope): Uint8Array {
  validateEnvelope(envelope);
  const body = new TextEncoder().encode(
    `{"v":${envelope.v},"id":${envelope.id},"type":${JSON.stringify(envelope.type)},"payload":${canonicalJSON(envelope.payload)}}`,
  );
  if (body.length > MAX_FRAME_BYTES) {
    throw new FramingError(
      "FRAME_TOO_LARGE",
      "encode",
      `length ${body.length} exceeds ${MAX_FRAME_BYTES}`,
    );
  }
  const frame = new Uint8Array(FRAME_HEADER_BYTES + body.length);
  new DataView(frame.buffer).setUint32(0, body.length, false);
  frame.set(body, FRAME_HEADER_BYTES);
  return frame;
}

function decodeBody(body: Uint8Array): Envelope {
  let parsed: unknown;
  try {
    parsed = JSON.parse(
      new TextDecoder("utf-8", { fatal: true }).decode(body),
    ) as unknown;
  } catch (error: unknown) {
    throw new FramingError(
      "INVALID_PAYLOAD",
      "decode",
      error instanceof Error ? error.message : "body is not valid JSON",
    );
  }
  if (!isPlainObject(parsed)) {
    throw new FramingError(
      "INVALID_PAYLOAD",
      "decode",
      "frame body is not a JSON object",
    );
  }
  if (parsed.v !== PROTOCOL_VERSION) {
    throw new FramingError(
      "VERSION_MISMATCH",
      "decode",
      `protocol version ${String(parsed.v)} is not supported`,
    );
  }
  const envelope: Envelope = {
    v: parsed.v as number,
    id: typeof parsed.id === "number" ? parsed.id : Number.NaN,
    type: typeof parsed.type === "string" ? parsed.type : "",
    payload: parsed.payload,
  };
  try {
    validateEnvelope(envelope);
  } catch (error: unknown) {
    if (error instanceof FramingError) {
      throw new FramingError("INVALID_PAYLOAD", "decode", error.message);
    }
    throw error;
  }
  return envelope;
}

/**
 * Incremental frame decoder. Bytes are pushed as they arrive and complete
 * frames are pulled out; a partial frame simply yields nothing until the rest
 * of it arrives.
 */
export class FrameDecoder {
  private pending: Uint8Array = new Uint8Array(0);
  private ended = false;

  /** Appends transport bytes to the decode buffer. */
  push(chunk: Uint8Array): void {
    if (chunk.length === 0) {
      return;
    }
    const combined = new Uint8Array(this.pending.length + chunk.length);
    combined.set(this.pending, 0);
    combined.set(chunk, this.pending.length);
    this.pending = combined;
  }

  /** Signals end of input. A partial frame then becomes a truncation. */
  end(): void {
    this.ended = true;
  }

  /** Reports bytes buffered but not yet forming a complete frame. */
  get buffered(): number {
    return this.pending.length;
  }

  /**
   * Returns the next complete frame, or undefined when more bytes are needed.
   * Throws FramingError for any protocol violation.
   */
  next(): Envelope | undefined {
    if (this.pending.length < FRAME_HEADER_BYTES) {
      if (this.ended && this.pending.length !== 0) {
        throw new FramingError(
          "FRAME_TRUNCATED",
          "read header",
          "incomplete length prefix",
        );
      }
      return undefined;
    }

    const view = new DataView(
      this.pending.buffer,
      this.pending.byteOffset,
      this.pending.byteLength,
    );
    const length = view.getUint32(0, false);
    if (length === 0) {
      throw new FramingError("FRAME_EMPTY", "read header");
    }
    // The length is validated before any allocation, so a hostile prefix cannot
    // make the process reserve 4 GiB.
    if (length > MAX_FRAME_BYTES) {
      throw new FramingError(
        "FRAME_TOO_LARGE",
        "read header",
        `length ${length} exceeds ${MAX_FRAME_BYTES}`,
      );
    }

    const total = FRAME_HEADER_BYTES + length;
    if (this.pending.length < total) {
      if (this.ended) {
        throw new FramingError(
          "FRAME_TRUNCATED",
          "read body",
          `announced ${length} bytes, received ${this.pending.length - FRAME_HEADER_BYTES}`,
        );
      }
      return undefined;
    }

    const body = this.pending.subarray(FRAME_HEADER_BYTES, total);
    // The frame boundary is honoured before decoding, so an invalid body leaves
    // the stream aligned and the session usable.
    this.pending = this.pending.subarray(total);
    return decodeBody(body);
  }
}

/** Decodes exactly one frame from a complete buffer. */
export function decodeFrame(frame: Uint8Array): Envelope {
  const decoder = new FrameDecoder();
  decoder.push(frame);
  decoder.end();
  const envelope = decoder.next();
  if (envelope === undefined) {
    throw new FramingError(
      "FRAME_TRUNCATED",
      "read",
      "buffer does not contain a whole frame",
    );
  }
  return envelope;
}

/** Streams envelopes from a chunked transport such as process.stdin. */
export async function* readFrames(
  source: AsyncIterable<Uint8Array | string>,
): AsyncGenerator<Envelope> {
  const decoder = new FrameDecoder();
  for await (const chunk of source) {
    decoder.push(
      typeof chunk === "string" ? new TextEncoder().encode(chunk) : chunk,
    );
    for (;;) {
      const envelope = decoder.next();
      if (envelope === undefined) {
        break;
      }
      yield envelope;
    }
  }
  decoder.end();
  for (;;) {
    const envelope = decoder.next();
    if (envelope === undefined) {
      return;
    }
    yield envelope;
  }
}
